network/p2p/p2pbuilder/gossipsub/gossipSubBuilder.go

```diff
@@ -0,0 +1,231 @@
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/network/channels"
+	"github.com/onflow/flow-go/network/p2p"
+	"github.com/onflow/flow-go/network/p2p/p2pnode"
+	"github.com/onflow/flow-go/network/p2p/scoring"
+	"github.com/onflow/flow-go/network/p2p/tracer"
+	"github.com/onflow/flow-go/network/p2p/utils"
+)
+
+// The Builder struct is used to configure and create a new GossipSub pubsub system.
+type Builder struct {
+	logger                       zerolog.Logger
+	metrics                      module.GossipSubMetrics
+	h                            host.Host
+	subscriptionFilter           pubsub.SubscriptionFilter
+	gossipSubFactory             p2p.GossipSubFactoryFunc
+	gossipSubConfigFunc          p2p.GossipSubAdapterConfigFunc
+	gossipSubPeerScoring         bool          // whether to enable gossipsub peer scoring
+	gossipSubScoreTracerInterval time.Duration // the interval at which the gossipsub score tracer logs the peer scores.
+	// gossipSubTracer is a callback interface that is called by the gossipsub implementation upon
+	// certain events. Currently, we use it to log and observe the local mesh of the node.
+	gossipSubTracer             p2p.PubSubTracer
+	peerScoringParameterOptions []scoring.PeerScoreParamsOption
+	idProvider                  module.IdentityProvider
+	rsys                        routing.Routing
+}
+
+var _ p2p.GossipSubBuilder = (*Builder)(nil)
+
+// SetHost sets the host of the builder.
+// If the host has already been set, a fatal error is logged.
+func (g *Builder) SetHost(h host.Host) {
+	if g.h != nil {
+		g.logger.Fatal().Msg("host has already been set")
+		return
+	}
+	g.h = h
+}
+
+// SetSubscriptionFilter sets the subscription filter of the builder.
+// If the subscription filter has already been set, a fatal error is logged.
+func (g *Builder) SetSubscriptionFilter(subscriptionFilter pubsub.SubscriptionFilter) {
+	if g.subscriptionFilter != nil {
+		g.logger.Fatal().Msg("subscription filter has already been set")
+	}
+	g.subscriptionFilter = subscriptionFilter
+}
+
+// SetGossipSubFactory sets the gossipsub factory of the builder.
+// We expect the node to initialize with a default gossipsub factory. Hence, this function overrides the default config.
+func (g *Builder) SetGossipSubFactory(gossipSubFactory p2p.GossipSubFactoryFunc) {
+	if g.gossipSubFactory != nil {
+		g.logger.Warn().Msg("gossipsub factory has already been set, overriding the previous factory.")
+	}
+	g.gossipSubFactory = gossipSubFactory
+}
+
+// SetGossipSubConfigFunc sets the gossipsub config function of the builder.
+// We expect the node to initialize with a default gossipsub config. Hence, this function overrides the default config.
+func (g *Builder) SetGossipSubConfigFunc(gossipSubConfigFunc p2p.GossipSubAdapterConfigFunc) {
+	if g.gossipSubConfigFunc != nil {
+		g.logger.Warn().Msg("gossipsub config function has already been set, overriding the previous config function.")
```

Comment:
same here

Reply:
At present, the overriding of the gossipsub factory is primarily used for testing. It is expected that this will seldom be done in production environments. Logging at the warn level can help identify such cases, making it easier to spot when the gossipsub factory has been overridden outside of testing scenarios.